packages/server-wallet/src/objectives/objective-manager.ts

```diff
@@ -0,0 +1,197 @@
+            this.store,
+            objective.data.targetChannelId,
+            tx
+          );
+          executeProtocol = () =>
+            OpenChannelProtocol.protocol(protocolState as OpenChannelProtocol.ProtocolState);
+        } else if (objective.type === 'CloseChannel') {
+          protocolState = await CloseChannelProtocol.getCloseChannelProtocolState(
+            this.store,
+            objective.data.targetChannelId,
+            tx
+          );
+          executeProtocol = () =>
+            CloseChannelProtocol.protocol(protocolState as CloseChannelProtocol.ProtocolState);
+        } else {
+          throw new Error('Unexpected objective');
+        }
+
+        const doAction = async (action: ProtocolAction): Promise<any> => {
+          switch (action.type) {
+            case 'SignState': {
+              const {myIndex, participants, channelId} = protocolState.app;
+              const signedState = await this.store.signState(action.channelId, action, tx);
+              createOutboxFor(channelId, myIndex, participants, {
+                signedStates: [signedState],
+              }).map(outgoing => outbox.push(outgoing));
+              return;
+            }
+            case 'FundChannel':
+              await this.store.addChainServiceRequest(action.channelId, 'fund', tx);
+              await this.chainService.fundChannel({
+                ...action,
+                expectedHeld: BN.from(action.expectedHeld),
+                amount: BN.from(action.amount),
+              });
+              return;
+            case 'CompleteObjective':
+              await this.store.markObjectiveAsSucceeded(objective, tx);
+              channelResults.push(ChannelState.toChannelResult(protocolState.app));
+              attemptAnotherProtocolStep = false;
+              return;
+            case 'Withdraw':
+              await this.store.addChainServiceRequest(action.channelId, 'withdraw', tx);
+              // app.supported is defined (if the wallet is functioning correctly), but the compiler is not aware of that
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+              await this.chainService.concludeAndWithdraw([protocolState.app.supported!]);
+              return;
+            case 'RequestLedgerFunding': {
+              const ledgerChannelId = await determineWhichLedgerToUse(protocolState.app, tx);
+              await LedgerRequest.requestLedgerFunding(
+                protocolState.app.channelId,
+                ledgerChannelId,
+                tx
+              );
+              return;
+            }
+            case 'RequestLedgerDefunding': {
+              const ledgerChannelId = await determineWhichLedgerToUse(protocolState.app, tx);
+              await LedgerRequest.requestLedgerDefunding(
+                protocolState.app.channelId,
+                ledgerChannelId,
+                tx
+              );
+              return;
+            }
```

Comment:
If we're in a refactoring mood, I feel like these could be combined into one

Reply:
I was intending to move this down into an objective model shortly.

Reply:
I would advise against putting business logic in the model, since it couples data access to business logic, making it hard to eg. Use a prepared statement rather than an objection model to make an update.